### PR TITLE
Add more tests for ImmutableInterlocked.InterlockedCompareExchangeArray

### LINKED ISSUE #17108: ImmutableInterlockedTests.InterlockedCompareExchangeArrayDefault has no assertions
The [`ImmutableInterlockedTests.InterlockedCompareExchangeArrayDefault`](https://github.com/dotnet/corefx/blob/343561048c85321713c30b6e69ddd20cfac94dea/src/System.Collections.Immutable/tests/ImmutableInterlockedTests.cs#L172) test is missing assertions, so while it provides some code coverage it's not currently able to reveal bugs aside from exceptions. Assertions regarding `oldValue` should probably be added to improve the quality of this test.

### PR DESCRIPTION
**Pull request by hughbe**
Also Fixes #17108

### CODE REVIEW

**Comment by stephentoub** [comment]
@dotnet-bot test Innerloop Windows_NT Release x64 Build and Test please (access denied failure on a temp file)

File: src/System.Collections.Immutable/tests/ImmutableInterlockedTests.cs [hunk 1]
```diff
@@ -169,22 +169,47 @@ public void InterlockedExchangeArrayNonDefault()
...
         {
             ImmutableArray<int> array = default(ImmutableArray<int>);
-            var oldValue = ImmutableInterlocked.InterlockedCompareExchange(ref array, ImmutableArray.Create<int>(4, 5, 6), default(ImmutableArray<int>));
+            var oldValue = ImmutableInterlocked.InterlockedCompareExchange(ref array, ImmutableArray.Create<int>(4, 5, 6), array);
+            Assert.Equal(default(ImmutableArray<int>), oldValue);
+            Assert.Equal(new[] { 4, 5, 6 }, array);
+        }
+
+        [Fact]
+        public void InterlockedCompareExchangeArray_LocationDefault_NotEqualToComarand()
```
Comment (stephentoub) [inline src/System.Collections.Immutable/tests/ImmutableInterlockedTests.cs]: Nit: Comarand => Comparand

File: src/System.Collections.Immutable/tests/ImmutableInterlockedTests.cs [hunk 2]
```diff
@@ -169,22 +169,47 @@ public void InterlockedExchangeArrayNonDefault()
         }
 
         [Fact]
-        public void InterlockedCompareExchangeArrayDefault()
+        public void InterlockedCompareExchangeArray_LocationDefault_EqualToComparand()
         {
             ImmutableArray<int> array = default(ImmutableArray<int>);
-            var oldValue = ImmutableInterlocked.InterlockedCompareExchange(ref array, ImmutableArray.Create<int>(4, 5, 6), default(ImmutableArray<int>));
+            var oldValue = ImmutableInterlocked.InterlockedCompareExchange(ref array, ImmutableArray.Create<int>(4, 5, 6), array);
+            Assert.Equal(default(ImmutableArray<int>), oldValue);
```
Comment (stephentoub) [inline src/System.Collections.Immutable/tests/ImmutableInterlockedTests.cs]: Nit: it looks like other tests use:
```C#
Assert.True(oldValue.IsDefault);
```
for this

File: src/System.Collections.Immutable/tests/ImmutableInterlockedTests.cs [hunk 3]
```diff
@@ -169,22 +169,47 @@ public void InterlockedExchangeArrayNonDefault()
...
 
-            var arrayBefore = array;
-            oldValue = ImmutableInterlocked.InterlockedCompareExchange(ref array, ImmutableArray.Create<int>(4, 5, 6), array);
-            Assert.Equal(oldValue.AsEnumerable(), arrayBefore);
+        [Fact]
+        public void InterlockedCompareExchangeArray_LocationNotDefault_NotEqualToNonDefaultComparand()
+        {
+            ImmutableArray<int> array = ImmutableArray.Create(1, 2, 3);
+            var oldValue = ImmutableInterlocked.InterlockedCompareExchange(ref array, ImmutableArray.Create<int>(4, 5, 6), array);
+            Assert.Equal(new[] { 1, 2, 3 }, oldValue);
```
Comment (stephentoub) [inline src/System.Collections.Immutable/tests/ImmutableInterlockedTests.cs]: Is there already a test that verifies the behavior when the array references are different but contain the same values?

Comment (hughbe) [inline src/System.Collections.Immutable/tests/ImmutableInterlockedTests.cs]: I've just added one

File: src/System.Collections.Immutable/tests/ImmutableInterlockedTests.cs (hunk 1 above)
Comment (hughbe) [inline src/System.Collections.Immutable/tests/ImmutableInterlockedTests.cs]: fixed, cheers

File: src/System.Collections.Immutable/tests/ImmutableInterlockedTests.cs (hunk 2 above)
Comment (hughbe) [inline src/System.Collections.Immutable/tests/ImmutableInterlockedTests.cs]: Addressed this

**Pull request merged**